pkg/expression/builtin_time.go

```diff
@@ -6233,11 +6233,28 @@ func addUnitToTime(unit string, t time.Time, v float64) (time.Time, bool, error)
 		}
 		tb = t.AddDate(0, int(v), 0)
 
+		originMonth := int(t.Month())
+		absV := int(math.Abs(v))
+		for originMonth < absV {
+			originMonth += 12
+		}
+
+		expectMonth := (originMonth + int(v)) % 12
+
+		loopCount := 0
+
 		// For corner case: timestampadd(month,1,date '2024-01-31') = "2024-02-29", timestampadd(month,1,date '2024-01-30') = "2024-02-29"
 		// `tb.Month()` refers to the actual result, `t.Month()+v` refers to the expect result.
 		// Actual result may be greater than expect result, we need to judge and modify it.
-		for int(tb.Month())%12 != (int(t.Month())+int(v))%12 {
+		for int(tb.Month())%12 != expectMonth {
 			tb = tb.AddDate(0, 0, -1)
+
+			// TODO remove it in the future
+			// Prevent infinity loop when there are bugs
+			loopCount++
+			if loopCount > 100000 {
+				panic("Enter into infinity loop because of bugs")
+			}
```

Comment:
It's unacceptable to panic here

Reply:
Date_add has the exactly problem and this pr(https://github.com/pingcap/tidb/pull/8988/files) introduce a new function
`types.AddDate` to fix the difference between mysql and golang api, I think you can just reuse that function instead of writing a new one.

Reply:
done